core/services/keystore/models.go

```diff
@@ -423,3 +424,68 @@ func (rawKeys rawKeyRing) keys() (*keyRing, error) {
 func adulteratedPassword(password string) string {
 	return "master-password-" + password
 }
+
+type ResourceMutex struct {
+	mu          sync.Mutex
+	activeCount map[ServiceType]int // Tracks active users per service type
+}
+type ServiceType int
+
+const (
+	TXMv1 ServiceType = iota
+	TXMv2
+)
+
+// TryLock attempts to lock the resource for the specified service type.
+// It returns an error if the resource is locked by a different service type.
+func (rm *ResourceMutex) TryLock(serviceType ServiceType) error {
+	rm.mu.Lock()
+	defer rm.mu.Unlock()
+
+	// Check if other service types are using the resource
+	for otherServiceType, count := range rm.activeCount {
+		if otherServiceType != serviceType && count > 0 {
+			return errors.New("resource is locked by another service type")
```

Comment:
If it is always incorrect to lock from a second service type, then why have a _map_ of counts? How about just a field for each?
```suggestion
		if rm.serviceType != serviceType && count > 0 {
			return errors.New("resource is locked by another service type")
```

Reply:
Yeah, that makes sense, changed in [`759f3e1` (#15969)](https://github.com/smartcontractkit/chainlink/pull/15969/commits/759f3e19731a3dfb4a3b74f81686313c508caabd)